pkg/util/tracingservice/client.go

```diff
@@ -0,0 +1,173 @@
+
+// localTraceClient executes the local tracing service's code to fetch inflight
+// span recordings.
+type localTraceClient struct {
+	tracer *tracing.Tracer
+}
+
+// newLocalTraceClient returns a local tracing service client.
+func newLocalTraceClient(tracer *tracing.Tracer) TraceClient {
+	return &localTraceClient{tracer: tracer}
+}
+
+// GetSpanRecordings implements the TraceClient interface.
+func (c *localTraceClient) GetSpanRecordings(
+	_ context.Context, traceID uint64,
+) (*tracingservicepb.SpanRecordingResponse, error) {
+	var resp tracingservicepb.SpanRecordingResponse
+	err := c.tracer.VisitSpans(func(span *tracing.Span) error {
+		if span.TraceID() != traceID {
+			return nil
+		}
+		for _, rec := range span.GetRecording() {
+			resp.SpanRecordings = append(resp.SpanRecordings, rec)
+		}
+		return nil
+	})
+	return &resp, err
+}
+
+var _ TraceClient = &remoteTraceClient{}
+
+// remoteClient uses the node dialer and tracing service clients to fetch
+// span recordings from other nodes.
+type remoteTraceClient struct {
+	traceClient tracingservicepb.TraceClient
+}
+
+// newRemoteClient instantiates a remote tracing service client.
+func newRemoteTraceClient(traceClient tracingservicepb.TraceClient) TraceClient {
+	return &remoteTraceClient{traceClient: traceClient}
+}
+
+// GetSpanRecordings implements the TraceClient interface.
+func (c *remoteTraceClient) GetSpanRecordings(
+	ctx context.Context, traceID uint64,
+) (*tracingservicepb.SpanRecordingResponse, error) {
+	return c.traceClient.GetSpanRecordings(ctx, &tracingservicepb.SpanRecordingRequest{TraceID: traceID})
+}
+
+// TraceClientDialer can be used to extract recordings from inflight spans for a
+// given traceID, from all nodes of the cluster.
+type TraceClientDialer struct {
+	tracer       *tracing.Tracer
+	localNodeID  roachpb.NodeID
+	dialer       *nodedialer.Dialer
+	nodeliveness migrationcluster.NodeLiveness
+}
+
+// NewTraceClientDialer returns a TraceClientDialer.
+func NewTraceClientDialer(
+	localNodeID roachpb.NodeID,
+	dialer *nodedialer.Dialer,
+	nodeliveness migrationcluster.NodeLiveness,
+	tracer *tracing.Tracer,
+) *TraceClientDialer {
+	return &TraceClientDialer{
+		localNodeID:  localNodeID,
+		dialer:       dialer,
+		nodeliveness: nodeliveness,
+		tracer:       tracer,
+	}
+}
+
+// GetSpanRecordingsFromCluster returns the inflight span recordings from all
+// nodes in the cluster.
+func (t *TraceClientDialer) GetSpanRecordingsFromCluster(
+	ctx context.Context, traceID uint64,
+) ([]tracingpb.RecordedSpan, error) {
+	var res []tracingpb.RecordedSpan
+	ns, err := migrationcluster.NodesFromNodeLiveness(ctx, t.nodeliveness)
+	if err != nil {
+		return res, err
+	}
+
+	// Collect spans from the local client.
+	localClient := newLocalTraceClient(t.tracer)
+	localSpanRecordings, err := localClient.GetSpanRecordings(ctx, traceID)
+	if err != nil {
+		return res, err
+	}
+	var mu syncutil.Mutex
+	res = append(res, localSpanRecordings.SpanRecordings...)
```

Comment:
Is this a pattern we use elsewhere? Sidestepping an RPC to talk to the local service directly? Given it's just observability, I think it's fine to not distinguish the local node and issue RPCs everywhere. I think gRPC is smart enough to short circuit the local request.

Reply:
Yeah, we use this in the blob service for nodelocal, but I agree probably not worth the additional complexity for observability.